extensions/wikia/LanguageWikisIndex/LanguageWikisIndexHooks.class.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+class LanguageWikisIndexHooks {
+
+	const WIKIS_INDEX_PAGE = '/language-wikis';
+
+	/**
+	 * Extension setup function, called on every request
+	 */
+	public static function onExtensionFunctions() {
+		if ( !self::isEmptyDomainWithLanguageWikis() ) {
+			return;
+		}
+		global $wgRequest;
+
+		$requestUrl = $wgRequest->getRequestURL();
+
+		// allow some other extensions (like robots) to render their content instead
+		if ( !Hooks::run( 'ShowLanguageWikisIndex', [ $requestUrl ] ) ) {
+			return;
+		}
+
+		self::handleRequest( $requestUrl, true );
+		exit( 0 );
+	}
+
+	public static function onGenerateRobotsRules() {
+		// don't render rules for an empty domain root
+		return !self::isEmptyDomainWithLanguageWikis();
+	}
+
+	public static function onClosedWikiPage( $requestUrl ) {
+		// when the English wiki is closed, we still want to render the list of existing language wikis
+		if ( count( WikiFactory::getLanguageWikis() ) > 0 && self::handleRequest( $requestUrl ) ) {
+			// request was handled by the handleRequest method, terminate
+			exit( 0 );
+		}
+		return true;	// proceed with rendering the closed wiki page
+	}
+
+	public static function isEmptyDomainWithLanguageWikis() {
+		global $wgCityId;
+		// we recognize an empty domain root with orphaned language path wikis by "fake" city id set by
+		// WikiFactoryLoader
+		return $wgCityId == WikiFactory::LANGUAGE_WIKIS_INDEX;
+	}
+
+	/**
+	 * Handles request on empty/closed wikis. For domain root, it redirects to the index page location. On the index
+	 * page, it renders the special page with the list of language wikis.
+	 *
+	 * @param $requestUrl
+	 * @param $redirectAll if true, all unknown urls will be redirected to index page. in that case true is returned.
+	 * @return bool True when the requestUrl was recognized and correct http response was sent
+	 */
+	private static function handleRequest( $requestUrl, $redirectAll = false ) {
+		global $wgTitle, $wgOut, $wgRequest, $wgSuppressCommunityHeader, $wgSuppressPageHeader;
+
+		switch ( parse_url( $requestUrl, PHP_URL_PATH ) ) {
+			case '/':
+				// use 302, maybe at some point an English wiki gets created
+				$wgRequest->response()->header( 'Location: ' . self::WIKIS_INDEX_PAGE , 302 );
+				return true;
+			case self::WIKIS_INDEX_PAGE:
+				// render the Special::LanguageWikisIndex page
+				$wgOut->disallowUserJs();
+
+				$wgTitle = SpecialPage::getTitleFor( 'LanguageWikisIndex' );
+
+				$context = $wgOut->getContext();
+
+				$context->setTitle( $wgTitle );
+				$context->setSkin( Skin::newFromKey( 'oasis' ) );
+				$wgSuppressCommunityHeader = true;
+				$wgSuppressPageHeader = true;
+
+				SpecialPageFactory::executePath( $wgTitle, $context );
+				$wgOut->output();
+				return true;
+			default:
+				if ( $redirectAll ) {
+					$wgRequest->response()->header( 'Location: ' . self::WIKIS_INDEX_PAGE , 302 );
```

Comment:
We should probably check with Becky about these being 302s. Even if the wiki is (re)created in the future, we may not do that any time soon and so 301s might be better to avoid junk sticking around in Google. Any thoughts @rogatty?

Reply:
Can we tell if it's a closed wiki or not? For closed wikis I'd use 301, for not-yet-created wikis I think 302 is fine. After all, we're returning 200 (come back later) on https://community.wikia.com/wiki/Community_Central:Not_a_valid_community?from=szl.starwars.wikia.com, not 410 (never come back). But yeah, I'd double check with Becky.